src/main/scala/mesosphere/marathon/core/instance/update/InstanceUpdater.scala

```diff
@@ -42,26 +51,38 @@ object InstanceUpdater extends StrictLogging {
   private def shouldBeExpunged(instance: Instance): Boolean =
     instance.tasksMap.values.forall(t => t.isTerminal) && instance.state.goal == Goal.Decommissioned
 
+  private def shouldAbandonReservation(instance: Instance): Boolean = {
+
+    def allAreGoneByOperator = instance.tasksMap.values.iterator
+      .flatMap(_.status.mesosStatus)
+      .forall(_.getState == MesosProtos.TaskState.TASK_GONE_BY_OPERATOR)
+
+    instance.reservation.nonEmpty && allAreGoneByOperator
+  }
+
   private[marathon] def mesosUpdate(instance: Instance, op: MesosUpdate): InstanceUpdateEffect = {
     val now = op.now
     val taskId = Task.Id.parse(op.mesosStatus.getTaskId)
     instance.tasksMap.get(taskId).map { task =>
       val taskEffect = task.update(instance, op.condition, op.mesosStatus, now)
       taskEffect match {
         case TaskUpdateEffect.Update(updatedTask) =>
-          val updated: Instance = updatedInstance(instance, updatedTask, now)
+          val updated: Instance = applyTaskUpdate(instance, updatedTask, now)
           val events = eventsGenerator.events(updated, Some(updatedTask), now, previousState = Some(instance.state))
           if (shouldBeExpunged(updated)) {
             logger.info("Requesting to expunge {}, all tasks are terminal, instance has no reservation and is not Stopped", updated.instanceId)
             InstanceUpdateEffect.Expunge(updated, events)
+          } else if (shouldAbandonReservation(updated)) {
+            val withoutReservation = updated.copy(agentInfo = None, reservation = None, state = updated.state.copy(condition = Condition.Scheduled))
```

Comment:
Shouldn't we drop the `tasksMap` as well?

Reply:
we can't do that, otherwise we'll lose the incarnation count.

I already wrote it to Tim privately, I'll just restate it here - the need to keep the tasksmap here is just problematic, I am pretty nervous about it and I am afraid it will cause bugs when updates come again etc. My idea would be to expunge and re-implement this with stable ids when we have incarnation on instance level so we don't have to risk introducing bugs while keeping stable instance ids at all costs...

Reply:
Second what Alena said.

Further, by the logic that we aren't changing the ID, I see no reason why we should drop the tasks map. They're still a part of the instance. We don't drop the tasks map when restarting an instance in any other scenario, ephemeral or not. (we do, however, replace them, when we schedule new ones).